core/lib/Cardano/Address/Script.hs

```diff
@@ -328,7 +406,15 @@ prettyErrValidateScript = \case
     Malformed ->
         "Parsing of the script failed. The script should be composed of nested \
         \lists, and the verification keys should be either encoded as bech32."
-
+    InvalidTimelocks ->
+        "The list inside a script must contain at most two timelock conditions \
+        \and they cannot be contradictory."
+    DuplicateXPubs ->
+        "Teh cosigners in a script template must assume an unique extended public key."
+    UnknownCosigner ->
+        "The script must use a cosigner present in a script template."
+    NoCosigner ->
+        "The script template must have at least one cosigner defined."
```

Comment:
It feels somewhat wrong to have these two last cases as part of the same error type for script validation. May I suggest to have instead 2 sum-types, one with errors that concerns `Script KeyHash` and one with others that concerns `Script Cosigner`? 

It's maybe not easy to re-conciliate the implementation with `validateScriptWith` ... we would need to pull off type families and GADTs likely... which may not be worth it for something _that_ simple.

Reply:
not we have separate script and separate script template errors.  one of script template error is when script is wrong, but validation of key hashes and cosigners is decoupled. please refer to code